pdg/src/info.rs

```diff
@@ -0,0 +1,229 @@
+    }
+}
+
+fn node_does_neg_offset(n: &Node) -> bool {
+    match n.kind {
+        NodeKind::Offset(x) => x < 0,
+        _ => false,
+    }
+}
+
+fn add_children_to_vec(g: &Graph, parents: &HashSet<NodeId>, v: &mut Vec<NodeId>) {
+    for (possible_child_id, possible_child_node) in g.nodes.iter_enumerated() {
+        if let Some(srcidx) = possible_child_node.source {
+            if parents.contains(&srcidx) {
+                v.push(possible_child_id);
+            }
+        }
+    }
+}
+
+fn check_flows_to_node_kind(
+    g: &Graph,
+    n: &NodeId,
+    node_check: fn(&Node) -> bool,
+) -> Option<NodeId> {
+    let mut seen = HashSet::new();
+    let mut to_view = Vec::new();
+    to_view.push(*n);
+    while let Some(node_id_to_check) = to_view.pop() {
+        if !seen.contains(&node_id_to_check) {
+            seen.insert(node_id_to_check);
+            let node_to_check: &Node = g.nodes.get(node_id_to_check).unwrap();
+            if node_check(node_to_check) {
+                return Some(node_id_to_check);
+            } else {
+                add_children_to_vec(g, &seen, &mut to_view);
+            }
+        }
+    }
+    return None;
+}
+
+fn greatest_desc(g: &Graph, n: &NodeId) -> NodeId {
+    let mut desc_seen = HashSet::<NodeId>::new();
+    let mut to_view = Vec::new();
+    let mut greatest_index = n.index();
+    let mut greatest_node_idx: NodeId = n.clone();
+    to_view.push(*n);
+    while let Some(node_id_to_check) = to_view.pop() {
+        if !desc_seen.contains(&node_id_to_check) {
+            desc_seen.insert(node_id_to_check);
+            if node_id_to_check.index() > greatest_index {
+                greatest_index = node_id_to_check.index();
+                greatest_node_idx = node_id_to_check;
+            }
+            add_children_to_vec(g, &desc_seen, &mut to_view);
+        }
+    }
+    return greatest_node_idx;
+}
+
+///Finds the highest-up ancestor of the given node n in g which is reached through Copy, Field, and
+///Offset, and returns its index as well as the Fields through which it is reached, in order
+///(the final element is the Field closest to the returned idx)
+fn calc_lineage(g: &Graph, n: &NodeId) -> (NodeId, Vec<Field>) {
+    let mut lineage = Vec::new();
+    let mut n_idx = n.clone();
+    loop {
+        let node = g.nodes.get(n_idx).unwrap();
+        let parent = match node.source {
+            None => break,
+            Some(p) => p,
+        };
+        n_idx = match node.kind {
+            NodeKind::Offset(_) => parent,
+            NodeKind::Copy => parent,
+            NodeKind::Field(f) => {
+                lineage.push(f);
+                parent
+            }
+            _ => break,
+        };
+    }
+    return (n_idx, lineage);
+}
+
+///Looks for a node which proves that the given node n is not unique. If any is found, it's
+///immediately returned (no guarantee of which is returned if multiple violate the uniqueness conditions);
+///otherwise None is returned.
```

Comment:
```suggestion
/// Looks for a node which proves that the given node n is not unique. If any is found, it's
/// immediately returned (no guarantee of which is returned if multiple violate the uniqueness conditions);
/// otherwise None is returned.
```